instrumentation/grpc-1.5/src/main/java/io/opentelemetry/instrumentation/hypertrace/grpc/v1_5/Http2HeadersInstrumentationModule.java

```diff
@@ -0,0 +1,133 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.hypertrace.grpc.v1_5;
+
+import static net.bytebuddy.matcher.ElementMatchers.failSafe;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import io.grpc.Metadata;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.hypertrace.agent.core.HypertraceSemanticAttributes;
+
+@AutoService(InstrumentationModule.class)
+public class Http2HeadersInstrumentationModule extends InstrumentationModule {
+
+  public Http2HeadersInstrumentationModule() {
+    super(GrpcInstrumentationName.PRIMARY, GrpcInstrumentationName.OTHER);
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return Arrays.asList(new NettyUtilsInstrumentation());
+  }
+
+  /**
+   * The server side HTTP2 headers are added in tracing gRPC interceptor. The headers are added to
+   * metadata in {@link GrpcUtils_convertHeaders_Advice}.
+   *
+   * <p>The client side HTTP2 headers are added directly to span in {@link
+   * Utils_convertClientHeaders_Advice}. TODO However it does not work for the first request
+   * https://github.com/hypertrace/javaagent/issues/109#issuecomment-740918018.
+   */
+  class NettyUtilsInstrumentation implements TypeInstrumentation {
+    @Override
+    public ElementMatcher<? super TypeDescription> typeMatcher() {
+      return failSafe(named("io.grpc.netty.Utils"));
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+      transformers.put(
+          isMethod().and(named("convertClientHeaders")).and(takesArguments(6)),
+          Utils_convertClientHeaders_Advice.class.getName());
+      transformers.put(
+          isMethod().and(named("convertHeaders")).and(takesArguments(1)),
```

Comment:
I think this is called multiple times from different flows in code.. How are we keeping track when to use and when not to use?

Reply:
correct we do not keep track of objects where should be added. I want to avoid using TLS and global map to do this as it is done in the old impl. 

The HTTP2 headers are added under different keys than the gRPC impl uses e.g. `:method.ht` to do not collide with user app. The implementation does not change any objects it just adds more data to metadata object passed around I think this is acceptable also the added headers are not added to the wire.

For completion: this advice does not change client response headers since those headers do not have the HTTP2 we are using (see the comment in the code). If the server side will cause any issues we could remove added headers in the interceptor before passing it to user define interceptors. The removal seems expensive so I would rather skip it for now.